script/urlChecker.ts

```diff
@@ -0,0 +1,93 @@
+import fetch from 'node-fetch';
+import ProgressBar from 'progress';
+import { argv } from 'yargs';
+import { content as contentRoute } from '../src/app/content/routes';
+import { Book, BookWithOSWebData, LinkedArchiveTreeSection } from '../src/app/content/types';
+import { findTreePages } from '../src/app/content/utils/archiveTreeUtils';
+import { getBookPageUrlAndParams, getUrlParamForPageId } from '../src/app/content/utils/urlUtils';
+import { assertDefined } from '../src/app/utils';
+import { findBooks } from './utils/bookUtils';
+
+const {
+  rootUrl,
+  bookId,
+  bookVersion,
+  archiveUrl,
+  useUnversionedUrls,
+} = argv as {
+  rootUrl?: string;
+  bookId?: string;
+  bookVersion?: string;
+  archiveUrl?: string;
+  useUnversionedUrls?: boolean;
+};
+
+async function checkPages(bookSlug: string, pages: string[]) {
+  let anyFailures = false;
+  const bar = new ProgressBar(`checking ${bookSlug} [:bar] :current/:total (:etas ETA)`, {
+    complete: '=',
+    incomplete: ' ',
+    total: pages.length,
+    width: 20,
+  });
+
+  const notFound: string[] = [];
+
+  for (const pageUrl of pages) {
+    try {
+      const response = await fetch(`${rootUrl}${pageUrl}`);
+      if (response.status === 404) {
+        notFound.push(pageUrl);
+      }
+    } catch (e) {
+      anyFailures = true;
+      bar.interrupt(`- (error loading) ${pageUrl}`);
+    }
+
+    bar.tick();
+  }
+
+  if (notFound.length) {
+    console.log(`404'd: ${notFound.length === pages.length // tslint:disable-line:no-console
+      ? 'all'
+      : `\n${notFound.join('\n')}\n`}`);
+  }
+
+  return anyFailures;
+}
+
+const getUrl = (book: Book) => useUnversionedUrls
+  ? (treeSection: LinkedArchiveTreeSection) =>
+      contentRoute.getUrl({
+        book: {
+          slug: (book as BookWithOSWebData).slug,
+        },
+        page: getUrlParamForPageId(book, treeSection.shortId),
+      })
+  : (treeSection: LinkedArchiveTreeSection) => getBookPageUrlAndParams(book, treeSection).url;
+
+async function checkUrls() {
+  const url = assertDefined(rootUrl, 'please define a rootUrl parameter, format: http://host:port');
+  const books = await findBooks({
+    archiveUrl,
+    bookId,
+    bookVersion,
+    rootUrl: url,
+  });
+
+  let anyFailures = false;
+
+  for (const book of books) {
+    const pages = findTreePages(book.tree);
+    anyFailures = await checkPages(book.slug, pages.map(getUrl(book)));
+  }
```

Comment:
This also could be speed up be using `Promise.all`

Reply:
I think doing so would be problematic with how the progress bars are printed. 

Besides that if lets say that each book has 200 pages and we chunk the pages by 50, books by 3.
I don't think there would be any benefit from running 50 requests for 3 books at the same time than running 150 requests for one book at the same time.

It would be slightly faster if the the last chunk would have only few pages in it, but I'm not sure if it's worth implementing.